src/oidcop/oidc/token.py

```diff
@@ -352,6 +363,253 @@ def post_parse_request(
+
+        try:
+            keyjar = _context.keyjar
+        except AttributeError:
+            keyjar = ""
+
+        try:
+            request.verify(keyjar=keyjar, opponent_id=client_id)
+        except (
+            MissingRequiredAttribute,
+            ValueError,
+            MissingRequiredValue,
+            JWKESTException,
+        ) as err:
+            return self.endpoint.error_cls(
+                error="invalid_request", error_description="%s" % err
+            )
+
+        _mngr = _context.session_manager
+        try:
+            _session_info = _mngr.get_session_info_by_token(
+                request["subject_token"], grant=True
+            )
+        except (KeyError, UnknownToken):
+            logger.error("Subject token invalid.")
+            return self.error_cls(
+                error="invalid_request",
+                error_description="Subject token invalid"
+            )
+
+        token = _mngr.find_token(_session_info["session_id"], request["subject_token"])
+
+        resp = self.enforce_policy(request, token)
+        if isinstance(resp, TokenErrorResponse):
+            return resp
+
+        if token.is_active() is False:
+            return self.error_cls(
+                error="invalid_request", error_description="Subject token inactive"
+            )
+        return resp
+
+    def enforce_policy(self, request, token):
+        _context = self.endpoint.server_get("endpoint_context")
+
+        if not isinstance(token, (AccessToken, RefreshToken)):
+            return self.error_cls(
+                error="invalid_request", error_description="Wrong token type"
+            )
+
+        if request["subject_token_type"] not in self.config["token_exchange"]["subject_token_types_supported"]:
+            return TokenErrorResponse(
+                error="invalid_request",
+                error_description="Unsupported subject token type",
+            )
+
+        if (
+            "requested_token_type" in request
+            and request["requested_token_type"] not in self.config["token_exchange"]["requested_token_types_supported"]
+        ):
+            return TokenErrorResponse(
+                error="invalid_request",
+                error_description="Unsupported requested token type",
+            )
+
+        if (
+            "requested_token_type" in request
+            and request["requested_token_type"] in self.config["token_exchange"]["policy"]
+        ):
+            fun = importer(self.config["token_exchange"]["policy"][request["requested_token_type"]]["callable"])
+            kwargs = self.config["token_exchange"]["policy"][request["requested_token_type"]]["kwargs"]
+        else:
+            fun = importer(self.config["token_exchange"]["policy"][""]["callable"])
+            kwargs = self.config["token_exchange"]["policy"][""]["kwargs"]
+
+        return fun(request, context=_context, subject_token=token, kwargs=kwargs)
```

Comment:
What happens if this raises an exception?

I think we should catch it and return a server error